crates/fluvio/src/producer/partition_producer.rs

```diff
@@ -217,12 +220,21 @@ where
 
         let (response, _) = self.send_to_socket(spu_socket, request).await?;
 
-        for (batch_notifier, partition_response_fut) in
-            batch_notifiers.into_iter().zip(response.into_iter())
+        for ((notify, batch_metadata), partition_response_fut) in batches_notify_and_metadata
+            .into_iter()
+            .zip(response.into_iter())
         {
-            if let Err(_e) = batch_notifier.send(partition_response_fut).await {
+            if let Err(_e) = notify.send(partition_response_fut.clone()).await {
                 trace!("Failed to notify produce result because receiver was dropped");
             }
+
+            if let Some(callback) = self.callback.clone() {
+                let event = ProduceCompletionEvent {
+                    created_at: batch_metadata.created_at,
+                    metadata: batch_metadata.clone(),
+                };
+                callback.finished(event).await.unwrap();
```

Comment:
this should be trigger when flush happen not wait callback is waiting.  this will block sender